### PR TITLE
export new precompute api in zebra-script

### LINKED ISSUE #1345: Reform data in <ScriptVerifier as Service>::Request
**Is your feature request related to a problem? Please describe.**

Because of limitations of the `zcash_script` API, the requests made to the script verification service have to include the entire transaction, not just the unlock script, outpoint, and any other auxiliary data.  This means we duplicate work between script verification results, and provide the script verifier with more data than it should require.

**Describe the solution you'd like**

In conjunction with https://github.com/zcash/zcash/issues/4876, modify the `zcash_script` API to take the minimal data needed for script verification, and then change the `Request` type of the `Service` impl to contain only that data.

### PR DESCRIPTION
**Pull request by yaahc**
## Motivation

The current version of zebra_script has an issue where it serializes and deserializes the same transaction for every input in that transaction, which can cause massive overhead when there are many inputs in a transaction.

## Solution

This PR leverages the new `PrecomputedTransaction` API @str4d exported in `libzcash_script`. We pass the serialized transaction to the c++ code once in order to get a pointer to the deserialized representation that the C++ code understands, then we pass that pointer back into the C++ code every time we verify an input.

The code in this pull request has:
  - [x] Documentation Comments
  - [x] Unit Tests and Property Tests

## Related Issues

Closes #1345

### CODE REVIEW

**Review by dconnolly — COMMENTED** [review]
Make clippy-stable happy and i'm happy to merge 👍

**Comment by yaahc** [comment]
> Make clippy-stable happy and i'm happy to merge +1

I'm gonna go ahead and clean things up a bit first, I want to remove the old API and update the rest of the codebase to use the new one before going ahead and merging.

File: zebra-script/src/lib.rs
Comment (yaahc) [inline zebra-script/src/lib.rs:111]: Removing this `fn` is a breaking change but because we almost certainly have nobody depending on the existing API and the fact that we've never published these crates I don't think we should worry about bumping the version number.

Comment (teor2345) [reply zebra-script/src/lib.rs:111]: I'm happy to go either way with this one.

Since we've tagged a release with the old version, it's good practice to bump the version for this change. But we didn't make any stability guarantees with that release, so maybe we don't need to do that.

Let's discuss at the Zebra sync?

File: zebra-script/src/lib.rs
Comment (teor2345) [inline zebra-script/src/lib.rs]: Can we also have a separate test that re-uses a pre-computed script?

I'd like a separate test because single-use and multi-use have different semantics.

File: zebra-script/src/lib.rs
Comment (teor2345) [inline zebra-script/src/lib.rs]: Can we also have a separate test that re-uses a pre-computed script?

Let's test the following input combinations:
* fail / fail
* fail / success
* success / fail

**Review by teor2345 — CHANGES_REQUESTED** [review]
I think we're missing some `zebra-script` tests here - the new API is more complex than the old API, so we need to test re-using pre-computed values.

Do we have any tests for the `zebra-consensus` part of this change?

We've merged bugs in this code before, so I'd like to make sure the post-Sapling sync test passes before we merge. @dconnolly do we still need to re-enable it? What's the ticket?

**Comment by teor2345** [comment]
Blocked on re-enabling post-Sapling sync tests.

**Comment by teor2345** [comment]
Does this PR close #1345?

**Comment by mpguerra** [comment]
> Blocked on re-enabling post-Sapling sync tests.

Is there an issue for this? I wasn't able to find it. I will create it if not.

**Comment by teor2345** [comment]
> > Blocked on re-enabling post-Sapling sync tests.
> 
> Is there an issue for this? I wasn't able to find it. I will create it if not.

I looked for one and couldn't find it, maybe @dconnolly knows?

**Comment by teor2345** [comment]
We decided to keep incrementing the minor version, until we get out of alpha.

> A pre-release version indicates that the version is unstable and might not satisfy the intended compatibility requirements as denoted by its associated normal version.

https://semver.org/

**Comment by yaahc** [comment]
> Does this PR close #1345?

yes

File: zebra-script/src/lib.rs
Comment (teor2345) [inline zebra-script/src/lib.rs]: Do we need a SAFETY comment here?

**Review by teor2345 — DISMISSED** [review]
This all looks good.

Should we add SAFETY comments for the new uses of `unsafe` in this PR?

**Comment by teor2345** [comment]
Unblocked by #1529, closing and re-opening to ensure those tests pass.

File: zebra-script/src/lib.rs
Comment (teor2345) [inline zebra-script/src/lib.rs]: Split into two sentences:
```suggestion
// Future + Send + Sync + static>`. In `zebra_consensus/src/transaction.rs`, an
// async block owns a `CachedFfiTransaction`, and holds it across an await
// point, while the transaction verifier is spawning all of the script verifier
// futures. The service readiness check requires this await between each task
// spawn. Each `script` future needs a copy of the
```

File: zebra-script/src/lib.rs
Comment (teor2345) [inline zebra-script/src/lib.rs]: This is a C enum issue. bindgen is correctly representing the implementation-defined behavior allowed by the C standard.

C enums don't have a guaranteed size:
> 6.7.2.2 Enumeration specifiers
> [...]
> Constraints
> The expression that defines the value of an enumeration constant shall be an integer constant expression that has a value representable as an int.
> [...]
> Each enumerated type shall be compatible with char, a signed integer type, or an unsigned integer type. The choice of type is implementation-defined, but shall be capable of representing the values of all the members of the enumeration.

http://www.open-std.org/JTC1/SC22/WG14/www/docs/n1256.pdf

If we want to avoid these kinds of conversions, we should avoid using enums in our C interfaces. Then bindgen will create Rust interfaces that don't require integer size conversions.

```suggestion
                flags.try_into().expect("zcash_script_SCRIPT_FLAGS_VERIFY_* enum values fit in a c_uint"),
```

**Review by teor2345 — COMMENTED** [review]
Just some minor suggestions, feel free to admin-merge after they're applied

File: zebra-script/src/lib.rs
Comment (teor2345) [inline zebra-script/src/lib.rs]: rustfmt
```suggestion
                flags
                    .try_into()
                    .expect("zcash_script_SCRIPT_FLAGS_VERIFY_* enum values fit in a c_uint"),
```

**Review by teor2345 — APPROVED** [review]
I just added a rustfmt commit, we should be good to merge after CI passes

**Pull request merged**
